### ISSUE TITLE
Manage dependencies

### ISSUE THREAD

**Issue by eric-wieser** [cc-scripts/cc-scripts#10]
Now that my turtletracker stuff is not part of the default install (which is not a problem), I feel we need some kind of dependency management.

For instance, I want to be able to do something like:

```
> ccs install program position
installing dependency "turtletracker"
    installing dependency "direction"
installing program "position"
> ccs install program direction
installing program "direction"
> ccs update program direction
updating dependency "turtletracker"
    updating dependency "direction"
updating program "direction"
```

I can think of a number of ways to work out which programs need which dependencies:
1. Use `ccs` itself to keep track of which programs require which dependcies
2. Parse out something like `loadAPI` from the lua code itself
3. Add a special comment at the top of the file which can be parsed.
4. Actually run the code, but abort as soon as a `loadAPI` is hit that can't be resolved, resolve it, and repeat.
5. Declare the code in a  way that makes running it safe:
   
   ``` lua
   cc_scripts.program(...) {
       requires = {'a', 'b', 'c'},
       usage = {"myprogram (x|y) <foo>", "myprogram <bar>"},
       function(args)
           if args.x then
               a.someAPICall(args['<foo>']
           end
       end
   }
   ```
6. Download APIs on the fly as they are needed (when the program runs for the first time). Not sure how this should handle updates.

**Comment by damien** [comment]
I've been thinking about this a bit since before you opened up this issue. I've been at a bit of a loss as to how to do this in a way that's not too hard to implement and that keeps the barrier of entry to contributing new cc-scripts programs low.

I like a lot of your ideas. In particular, I think code declaration you outlined in your 5th point makes a lot of sense and hits a sweet spot between ease of implementation and barrier of entry for people new to the codebase.

Are there any Lua projects that do something similar? Looking at LuaRocks docs, we could do something similar: http://www.luarocks.org/en/Creating_a_rock

**Comment by eric-wieser** [comment]
Point 5 is kinda doing two things at once - tracking requirements, but also implementing something like [docopt](docopt.org). The first half is far more trivial than the second.

Luarocks feels like overkill for what is essentially a bunch of internal packages

I'm warming to point 6.

**Comment by eric-wieser** [comment]
One way updates could be handled with an automatic downloader (point 6):
1. User runs `> ccs update position`
   - Downloads the new position file
   - Appends `programs/position` to `/cc-scripts/.updating`
2. User runs `> position`
   - Program hits `loadAPI('direction')`
     - _Somehow_, programs checks for themselves within `/cc-scripts/.updating` (can a script get its own name?). If present in that file:
       - download a new version of `direction`
       - Append `apis/direction` to `/cc-scripts/.updating`
       - Run `apis/direction`, repeating above steps recursively
       - Remove `apis/direction` from `/cc-scripts/.updating`
   - Program continues running
   - Program terminates - _Not sure this is possible to detect without modifying one of the program and the shell_
     - Remove `position` from `/cc-scripts/.updating`

**Comment by damien** [comment]
I agree LuaRocks is overkill; I was just using it as an example. I'm thinking all we need is a cookie-cutter, declarative way of specifying dependencies for any given program.

I'm also 100% behind something like docopt support. Coming from ruby, I know of a few projects that provide executables ways of generating nicely formatted usage instructions. See ruby's [optparse](http://ruby-doc.org/stdlib-1.9.3/libdoc/optparse/rdoc/OptionParser.html) and [mixlib-cli](http://rubydoc.info/gems/mixlib-cli/1.2.2/frames) for a couple examples.

As for updates--I'm thinking updates should be explicit for a couple reasons:
- Simplicity. Implementing an auto-update mechanism is another moving part we'd have to build, test, and probably break at some point.
- I think the `ccs` script is the best place to manage dependency resolution. Triggering it from elsewhere seems like it'd be more trouble than it's worth.

**Comment by eric-wieser** [comment]
One advantage of having the `loadAPI` call do the api download is that I can write a script to use APIs I haven't downloaded, and cc-scripts will sort it out for me - having dependencies applied by `ccs` means that I can only have script from the repository install their dependencies.

There's a much easier way to do point 6:
1. User runs > ccs update position
   - Downloads the new position file (if this isn't a user script)
   - Appends `programs/position` to `/cc-scripts/.updating`
2. User runs > position
   - `shell.run` is already overridden, and check if `position` is updating
     - If so, set a global flag (`__ccsForceRedownload`) in the environment of the program (so it goes away when the program ends/errors)
     - And remove  `programs/position` from `/cc-scripts/.updating`
     - Run the file normally
   - When loading APIs, download them if either they're not present, or said global flag is set

**Comment by eric-wieser** [comment]
We're now up to a state where the user should never have to install APIs themselves - This happens automagically when an api is loaded that does not exist. As a result, the bootstrapper now only installs the `cc_scripts` api, which is a core part of the bootstrapper itself.

API loading could probably do with some more careful error checking.

APIs will currently not autoupdate along with programs. To force an update, you can set the `_G.__ccsForceReload` global variable from a lua prompt before running the code. This also has to be cleared by hand (or a restart). Another option is to just delete the apis you want to upgrade.

Still a WIP.
